lib/linguist/heuristics.yml

```
@@ -686,7 +686,7 @@ disambiguations:
 - extensions: ['.stl']
   rules:
   - language: STL
-    pattern: '\A\s*solid(?=$|\s)(?:.|[\r\n])*?^endsolid(?:$|\s)'
+    pattern: '\A\s*solid[\s\S]*^endsolid(?:$|\s)'
```

1.	First of all, `solid[\s\S]` will match stuff like `solidobject` (which I presume isn't valid STL syntax), so either `(?=$|\s)` or `(?:$|\s)` is necessary here.

2.	Second, because this heuristic is anchored to the first-line of input (`\A`), it's possible to match the terminating `endsolid` directive using a separate regex:
	```suggestion
	    and:
	    - pattern: '\A\s*solid(?:$|\s)'
	    - pattern: '^\s*endsolid(?:$|\s)'
	```
	(**Nota bene:** I haven't tested these changes locally)

---

Is there a reason why `(?:$|\s)` is preferred over `\b` in this case?

---

> 2\. Second, because this heuristic is anchored to the first-line of input (`\A`), it's possible to match the terminating `endsolid` directive using a separate regex:

Splitting things makes this much more expensive. The second regex here is really expensive ([11144 steps](https://regex101.com/r/6vXA8K/1)) and quite possibly going to lead to another case of things running amok as the files grow compared my initial suggestion ([29 steps](https://regex101.com/r/VajTr7/1)), which now you point it out, needs improving for the situation you mentioned.

These all seem to match as be much more performant:

- `\A\s*solid(?:$|\s)[\s\S]*^endsolid(?:$|\s)` => [33 steps](https://regex101.com/r/HSQ919/1)
- `\A\s*solid\b[\s\S]*^endsolid(?:$|\s)` => [30 steps](https://regex101.com/r/i5LFeZ/1)
- `\A\s*solid\b[\s\S]*^endsolid\b` => [28 steps](https://regex101.com/r/aqC12C/1)

---

Sorry for the perf problems, everyone.

Does regex101 reliably approximate the cost to run these regexps in Ruby? The good original, before my bad version, was `'\A\s*solid(?=$|\s)(?m:.*?)\Rendsolid(?:$|\s)'`; the closest thing that works in regex101 is `/\A\s*solid(?=$|\s)(?:.*?)\nendsolid(?:$|\s)/gms`, and it claims that is super expensive ([107792 steps](https://regex101.com/r/HSQ919/1)).

---

OK, I tested the regexps in Ruby.

None of them perform badly on the test case, or on [a 5000 line file I found](https://github.com/prusa3d/PrusaSlicer/blob/c05542590d7c2d73eb69bbf7a82a482a075815c1/resources/profiles/Sovol/SV05_bed.stl). 

Here's how long each regex takes to run on the 500-line file, in milliseconds:

```
original: 1.9683429999859072
current:  2.581663000048138
proposed: 1.2971370000159368
```

regex101 is using a different regex engine.

---

The numbers above are for Ruby 3.2.2 on my laptop. Here's the code:

<details><summary>source code</summary>
```ruby
require 'benchmark'

ORIGINAL_RE  = /\A\s*solid(?=$|\s)(?m:.*?)\Rendsolid(?:$|\s)/
CURRENT_RE   = /\A\s*solid(?=$|\s)(?:.|[\r\n])*?^endsolid(?:$|\s)/
PROPOSED_RE  = /\A\s*solid[\s\S]*^endsolid(?:$|\s)/

text = File.read("SV05_bed.stl")

t = Benchmark.measure {
  1000.times do
    ORIGINAL_RE.match?(text)
  end
}
puts "original: #{t.real}"

t = Benchmark.measure {
  1000.times do
    CURRENT_RE.match?(text)
  end
}
puts "current:  #{t.real}"

t = Benchmark.measure {
  1000.times do
    PROPOSED_RE.match?(text)
  end
}
puts "proposed: #{t.real}"
```
<details>

---

> regex101 is using a different regex engine.

Ruby and PCRE differ in their interpretation of "multiline" modifiers (the parts scoped with `(?m:…)`). Specifically:

<table>
	<thead>
		<tr>
			<th></th>
			<th><h1><abbr title="multiline"><samp><b>(?m:</b><ins>...</ins><b>)</b></samp></abbr></h1> means…</th>
			<th><h1><abbr title="dotall"><samp><b>(?s:</b><ins>...</ins><b>)</b></samp></abbr></h1> means…</th>
		</tr>
	</thead>
	<tr>
		<th>Ruby</th>
		<td><code>.</code> matches any character, including newlines</td>
		<td>N/A (No such modifier)</td>
	</tr>
	<tr>
		<th>PCRE/Perl</th>
		<td><code>^</code> and <code>$</code> respectively match the beginning and end of each line, rather than the input.</td>
		<td><code>.</code> matches any character, including newlines; implies <code>(?m)</code>.</td>
	</tr>
</table>

See, the behaviour that `(?m)` normally enables in Perl/PCRE is instead the default in Ruby; the `\A` and `\Z`/`\z` assertions are used to match the start and end of the input string, respectively (which would normally be achieved using `^` and `$` without multiline mode).

Now, why is this relevant? Because benchmarks could be skewed in favour of whichever regex happens to fail first (if it won't match `.` because the dotall modifier is needed, then the regex engine logically has no reason to continue reading past the first newline).